### PR TITLE
Add torch.autograd.graph.node_creation_hook

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* __->__ #189443

Adds a public context manager that, while active, attaches a backward
pre-hook and/or post-hook to every autograd Node created within it. Both
hooks are optional and have the same signatures as
Node.register_prehook/register_hook. This is a creation-time dual of
saved_tensors_hooks aimed at the memory-attribution use case: capture
thread-local state (e.g. a memory-metadata region) at node creation and
restore it around the node's backward execution.

Rather than handing a freshly created Node to a user callback, we attach
the pre/post hooks directly. This is a more restrictive API, but it is
robust to when exactly the node is observed: the hooks only run at
backward time when the node is fully constructed, so there is no
sensitivity to next_edges/saved-tensor state being wired up yet. It also
lets us reuse the existing registerFunctionPreHook machinery (including
compiled-autograd support) and drop the abstract node-wrapping callback
and the firing-reentrancy guard that a node-argument API would require.

The context manager also takes an always_call flag (mirroring
torch.nn.Module.register_forward_hook) and offers a from_context_manager
convenience constructor; see below.

Design notes, in review order:

- aten/src/ATen/NodeCreationHooks.{h,cpp}: TLS stack of entries, each an
  optional (prehook, posthook) SafePyObject pair plus an always_call
  flag, mirroring SavedTensorDefaultHooksTLS. Lives in ATen so
  at::ThreadLocalState can snapshot it across thread boundaries onto
  engine worker threads. A process-global is_initialized latch (not
  thread_local) lets the empty() check on the hot path skip the
  thread_local read entirely when the feature is never used, matching
  saved_tensors' zero-overhead-when-unused guarantee.
- torch/csrc/autograd/node.h, function.cpp: fire_node_creation_hooks
  attaches the hooks at grad_fn-attachment time. A per-node fired flag
  makes multi-output nodes (which bind one shared grad_fn to each output)
  attach exactly once.
- torch/csrc/autograd/variable.cpp: impl::set_gradient_edge is the single
  funnel where a grad_fn becomes attached to an output tensor; it covers
  codegen ops, custom Functions (C++ and Python via _wrap_outputs),
  in-place rebase, and CopySlices. This naturally excludes AccumulateGrad
  (attached via a different field) and engine-internal nodes (GraphRoot,
  Identity), which are never bound to output tensors. Alternatives
  considered were the Node constructor and set_next_edges: both fire
  exactly once (so would drop the fired flag) but also fire for internal
  and leaf nodes, which is the wrong behavior; set_history has strictly
  worse coverage than set_gradient_edge. The lazy AsStridedBackward0
  regeneration in VariableHooks::grad_fn assigns grad_fn_ directly and so
  gets an explicit call.
- torch/csrc/autograd/engine.h, python_engine.{h,cpp}: libtorch_cpu
  cannot call into Python, so firing dispatches through a new
  Engine::attach_node_creation_hooks virtual overridden by PythonEngine
  (same pattern as get_default_saved_variable_hooks). The prehook reuses
  registerFunctionPreHook; the posthook is attached as a new
  PyFunctionSinglePostHook that wraps a single callable so it can carry
  the per-hook always_call flag (the shared-dict PyFunctionPostHook
  cannot).
- always_call: FunctionPostHook gains a should_run_on_error() virtual
  (default false; PyFunctionSinglePostHook returns always_call).
  engine.cpp's call_function wraps the node's backward in a try/catch;
  on exception it runs the opted-in post hooks with empty grad_inputs
  (call_post_hooks_on_error) before rethrowing, so state-restoring hooks
  still fire on error. A secondary exception from such a hook is swallowed
  with a warning so it does not mask the original error, matching
  nn.Module's always_call behavior.
- torch/autograd/graph.py: the public node_creation_hook context manager
  plus node_creation_hook.from_context_manager(cm_factory), a convenience
  constructor that enters a fresh context manager before each hooked
  node's backward and exits it after (with always_call=True). It keeps a
  thread-local LIFO stack of live context managers to handle nesting and
  engine worker threads. On the error path __exit__ receives
  (None, None, None) since the engine does not expose the exception to
  post hooks; this is documented.

Test Plan:

New tests in test/test_autograd.py cover: pre/post firing on in-context
nodes, prehook-only and posthook-only, multi-output nodes attaching once,
nesting (outermost first), custom Functions, in-place ops, no firing
outside the context (and AccumulateGrad not covered), propagation to
create_graph double-backward and checkpoint recomputation via
ThreadLocalState, pre-before-post ordering, always_call running the post
hook on backward error (and being skipped without it, and its exception
being swallowed), and from_context_manager enter/exit around each node
including on error and under double-backward. A test in
test/inductor/test_compiled_autograd.py covers hooks firing under
compiled autograd.

```
python test/test_autograd.py -k node_creation_hook
python test/test_autograd.py TestAutograd
python test/test_autograd.py TestMultithreadAutograd
python test/inductor/test_compiled_autograd.py TestCompiledAutograd.test_node_creation_hook
```

All pass (18 node_creation_hook tests eager; 402 TestAutograd; 13
TestMultithreadAutograd; 1 compiled-autograd).

Known limitation: under compiled autograd, hooks fire on the success
path, but the always_call error path relies on the eager engine, so a
posthook with always_call=True does not run if the compiled backward
raises. Documented in the node_creation_hook docstring.

Authored with Claude Code.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo

### CODE REVIEW

File: torch/autograd/graph.py [hunk 1]
```diff
@@ -456,6 +457,165 @@ def disable_saved_tensors_hooks(error_message: str) -> Generator[None, None, Non
             torch._C._autograd._saved_tensors_hooks_disable(maybe_prev_message)
 
 
+class node_creation_hook:
```
Comment (ezyang) [inline torch/autograd/graph.py:460]: Not sure if we should still call this `node_creation_hook` now lol

File: torch/_C/_autograd.pyi [hunk 2]
```diff
@@ -139,6 +139,12 @@ def _push_saved_tensors_default_hooks(
     unpack_hook: Callable[[Any], torch.Tensor],
 ) -> None: ...
 def _pop_saved_tensors_default_hooks() -> None: ...
+def _push_node_creation_hook(
+    prehook: Callable[..., Any] | None,
```
Comment (ezyang) [inline torch/_C/_autograd.pyi:143]: I vacillated on `prehook` versus `pre_hook` but prehook seems more in line with the module hook naming

File: torch/csrc/autograd/engine.cpp [hunk 3]
```diff
@@ -831,6 +831,29 @@ static variable_list call_post_hooks(
...
+    try {
+      variable_list empty_grad_inputs;
+      (*hook)(empty_grad_inputs, inputs);
+    } catch (std::exception& e) {
+      TORCH_WARN(
+          "Ignoring exception from node_creation_hook always_call post hook "
+          "raised while handling another error in ",
+          fn.name(),
+          ": ",
+          e.what());
```
Comment (ezyang) [inline torch/csrc/autograd/engine.cpp:852]: There is precedent for this

File: torch/autograd/graph.py (hunk 1 above)
Comment (soulitzer) [reply torch/autograd/graph.py:460]: Could it be two APIs for flexibility? Also a prepend= seems useful to make sure we fully straddle existing hooks?
```
with node_prehook(hook1, prepend=True), node_posthook(hook2, always_call=True, prepend=False):
    pass
```
- posthook doesn't match Node.register_hook, but maybe its fine, we can always add Node.register_posthook as an alias later.
- node_prehook rather than push_node_prehook because push rather than override is assumed? saved_tensor_hook overrides rather than stacks though.

Comment (ezyang) [reply torch/autograd/graph.py:460]: I thought about calling the posthook "hook" but... ugh, I didn't like it haha.

What is the proposed semantics for `prepend`? You just want to control the relative ordering to the existing hooks?

Comment (soulitzer) [reply torch/autograd/graph.py:460]: Yeah I also like posthook over hook

> You just want to control the relative ordering to the existing hooks?

If the users' existing prehook does some allocation, maybe it would be good to run our prehook before theirs to capture that (the default is hooks registered first are fired first).

Comment (ezyang) [reply torch/autograd/graph.py:460]: I'm not super excited about having an API that let's you decide if the hook goes in front/back, unless we have precedent for it / there's a specific use case for it. The "compositional" strategy seems clearly that each inner nesting of node creation hook should be "outside" the previous hooks.

Comment (soulitzer) [reply torch/autograd/graph.py:460]: Yeah I would agree that the two-part + prepend API is kind of complicated. I think I mostly just prefer the underlying implementation to be structured to have to separate stacks, since unlike saved tensor hooks the two hooks here aren't necessarily logically coupled. To make users lives easier, could we have a thin python layer that just does the right recipe.

(We do have precedent though!  https://docs.pytorch.org/docs/2.13/generated/torch.nn.Module.html#torch.nn.Module.register_forward_hook)

Comment (ezyang) [reply torch/autograd/graph.py:460]: I think the problem is that if you are doing context managers, they are in principle coupled, and it would be better to be lexical by default. One way to sidestep this discussion is to just only support context manager, but that also feels bad?

Comment (soulitzer) [reply torch/autograd/graph.py:460]: > I think the problem is that if you are doing context managers, they are in principle coupled, and it would be better to be lexical by default.

I think we can even start with keeping "two separate API" private, and since context manager is our main use case, that is the main one we recommend?

btw I realize there are two types of different interactions that I was confusing earlier:

- interactions with grad_fn.register_hooks
- interactions with other pairs of hooks where we want their relative firing order to be lexical

For the latter, I think "prepend" is precisely what we need:

- I can either "postpend prehook + prepend posthook" to have my pair of hooks to be innermost scope OR "prepend prehook + postpend posthook" to have my pair of hooks be the outermost scope.

- I also realize you also don't need two separate list holding prepend ones and postpend ones as its just equivalent to prepending and postpending to a single list lol


For the former case, it is more annoying, like if you want hooks to be always the outer most, you need to maintain a separate list so that you can call it after all the old-style hooks have fired.

Comment (ezyang) [reply torch/autograd/graph.py:460]: Ugh, what a mess. I agree we need to work out semantics with regards to preexisting hooks, and I guess actually I just haven't done the posthooks lexically at all? Ughhhh

**Comment by soulitzer** [comment]
Closing this since we're going with the other approach and I keep accidentally opening this one lol
